spec-0006/index.md

```diff
@@ -0,0 +1,68 @@
+---
+title: "SPEC 6 — Keys to the Castle"
+date: 2022-12-19
+author:
+  - "Stéfan van der Walt <stefanv@berkeley.edu>"
+  - "Jarrod Millman <millman@berkeley.edu>"
+discussion:
+endorsed-by:
+---
+
+## Description
+
+<!--
+Briefly and clearly describe the proposal.
+Explain the general need and the advantages of this specific proposal.
+If relevant, include examples of how the new functionality would be used,
+intended use-cases, and pseudo-code illustrating its use.
+-->
+
+Project developers often have to share secrets, such as server passwords and social media logins.
+Typically, such secrets are distributed among those who need them, without any centralized system for tracking the secrets or who has access to them.
+When a server needs to be accessed, there is often a scramble to find someone with credentials.
+This SPEC discusses the requirements for a system to distribute secrets, and provides an example implementation.
+
+## Implementation
+
+### Principles
+
+A system for distributing project secrets must have the following properties:
+
+- Secrets are stored encrypted in a central (remote) location.
+- It must be possible to grant access to the secrets to a select group of team members.
+- It must be possible to revoke future access to the secrets.
+- The system must not rely on closed source or commercial encryption facilities, that
+  can later disappear or be made unavailable.
+
+### Example
+
+The [NumPy password vault](https://github.com/scientific-python/vault-template) is an example of an implementation that satisfies the above principles.
+The secrets are stored, encrypted, in a public Git repository.
+The vault uses [gopass](https://github.com/gopasspw/gopass) to manage access via GPG keys.
+Each secret is encrypted using the public keys of all developers that should have access.
+If a developer's access is removed, the vault is re-encrypted so that that developer cannot read future copies of the repository (but secrets should be considered compromised and, thus, rotated).
+
+### Alternative implementations
+
+Unknown
```

Comment:
I was trying to figure out some [PyPI security stuff](https://discuss.python.org/t/pypi-security-work-multifactor-auth-progress-help-needed/1042/53) where 1password's "free team for open source" was recommended. It looks like it's quite popular with some large projects like: numfocus, jupyter, Python Software Foundation Infrastructure Team, and Rust.

https://github.com/1Password/1password-teams-open-source

I think a big advantage here is access controls, which I'm not sure you can get for free with bitwarden.

Reply:
oh thanks for sharing that! I even see Kali in the list. If that's not a good sign, I don't know what would be 😅

Reply:
Actually I just heard that NumFOCUS provide team accounts for affiliated projects.

Reply:
The NumFOCUS folks just point you to that 1Password page 😅 but yes, we are using 1Password for Teams for Open Source for napari and it was a [smooth process](https://github.com/1Password/1password-teams-open-source/pull/708).

Reply:
Oh really? I heard a different story from another project. That would be a bit underwhelming but ok still nice to have a central place saying what exists and is possible to do.

Reply:
@jni, could you add all the core devs to the team? It's been a little unclear to me if the team accounts were capped at 10.

Reply:
Well, not all 14 core devs have asked for invites so we haven't tested that hypothesis yet 😂 , but nothing in the process suggested that limit. They explicitly add for the number of core contributors in the PR template, so I presume that's the limit they set? But probably can be expanded?

Reply:
This is a great option; thanks for finding it! I've added it to the SPEC.

Reply:
1password's [pricing page](https://1password.com/business-pricing) says the "teams" account is up to ten people, and after that it's called "business".

But not sure what this translates to for the open source projects.